### PR TITLE
refactor(security): improve SecureRandom implementation, tests, and docs

### PR DESCRIPTION
**Pull request by joshtrichards**
## Summary

This PR improves the SecureRandom implementation along with its related tests and documentation.

### Key changes

- **Refactored implementation**:  
  - Added strict checks to prevent accidental use of non-ASCII, duplicate-containing, or overly short character sets.
    - This is a security-hardening change: it prevents developers from making subtle, security-impacting mistakes.
    - Previously, passing a character set shorter than 4 characters, containing multibyte characters, or with duplicates would (most likely) have resulted in unexpected, but not necessarily fatal, behavior. The function would still "work," but might return poor quality random strings, cause subtle bugs, or perhaps fail in different ways.
    - Technically a backward-incompatible (breaking) change, but unlikely to be disruptive in practice:
      - No known use cases intentionally violate these checks (any that do are likely unintentional, so this will surface them while preventing future occurrences).
  - Established a new constant containing the (existing) default `$characters` set passed to `generate()`.
  - Improved docblocks and in-code documentation for `ISecureRandom` and `SecureRandom` to better describe behavior, parameters, and usage.

- **Expanded and updated tests**:  
  - Added new tests for:
    - Invalid character sets (duplicates, too short, non-ASCII)
    - Unusual/oddball lengths (e.g., 31) to check for off-by-one and fencepost errors
    - Minimum viable character set size and large custom sets
    - Output uniqueness (basic sanity check)
    - Default and user-supplied character sets  
  - Improved existing test coverage and runtime performance:
    - Replaced impractically large test lengths (e.g., 64,000) with more practical ranges for CI, while still ensuring coverage of edge cases and real-world needs.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)
- [x] [Labels added](https://github.com/nextcloud/server/labels) where applicable (ex: bug/enhancement, `3. to review`, feature component)
- [x] [Milestone added](https://github.com/nextcloud/server/milestones) for target branch/version (ex: 32.x for `stable32`)

### CODE REVIEW

File: tests/lib/Security/SecureRandomTest.php [hunk 1]
```diff
@@ -81,4 +81,56 @@ public function testInvalidLengths($length): void {
...
+		return [
+			'invalid_too_short' => ['abc'],
+			'invalid_duplicates' => ['aabcd'],
+			'invalid_non_ascii' => ["abcd\xf0"],
+		];
+	}
+
+	/**
+	 * @dataProvider invalidCharProviders
+	 */
```
Comment (nickvergessen) [inline tests/lib/Security/SecureRandomTest.php]:
```suggestion
	#[DataProvider('invalidCharProviders')
```

(and add the class import)

File: tests/lib/Security/SecureRandomTest.php [hunk 2]
```diff
@@ -81,4 +81,56 @@ public function testInvalidLengths($length): void {
...
+	public function testInvalidCharacterSets(string $invalidCharset): void {
+		$this->expectException(\InvalidArgumentException::class);
+		$this->rng->generate(10, $invalidCharset);
+	}
+
+	public function testDefaultCharsetBase64Characters(): void {
+		$randomString = $this->rng->generate(100);
+		$this->assertMatchesRegularExpression('/^[A-Za-z0-9\+\/]+$/', $randomString);
```
Comment (nickvergessen) [inline tests/lib/Security/SecureRandomTest.php]:
```suggestion
		$this->assertMatchesRegularExpression('/^[A-Za-z0-9\+\/]{100}$/', $randomString);
```

File: tests/lib/Security/SecureRandomTest.php [hunk 3]
```diff
@@ -81,4 +81,56 @@ public function testInvalidLengths($length): void {
...
+		$first = $this->rng->generate(1000);
+		$second = $this->rng->generate(1000);
+		$this->assertNotEquals($first, $second, "Random output should not be repeated.");
+	}
+
+	public function testMinimumValidCharset(): void {
+		$charset = 'abcd';
+		$randomString = $this->rng->generate(500, $charset);
+		$this->assertMatchesRegularExpression('/^[abcd]+$/', $randomString);
+		$this->assertEquals(500, strlen($randomString));
```
Comment (nickvergessen) [inline tests/lib/Security/SecureRandomTest.php]:
```suggestion
		$this->assertMatchesRegularExpression('/^[abcd]{500}$/', $randomString);
```

Or keep the length check separate but then you can also do that upstairs.

File: tests/lib/Security/SecureRandomTest.php [hunk 4]
```diff
@@ -81,4 +81,56 @@ public function testInvalidLengths($length): void {
...
+		$randomString = $this->rng->generate(200, $charset);
+		foreach (str_split($randomString) as $char) {
+			$this->assertStringContainsString($char, $charset);
+		}
+	}
+
+	public function testUserProvidedValidCharset(): void {
+		$charset = '@#$!';
+		$randomString = $this->rng->generate(64, $charset);
+		$this->assertMatchesRegularExpression('/^[@#$!]+$/', $randomString);
```
Comment (nickvergessen) [inline tests/lib/Security/SecureRandomTest.php]:
```suggestion
		$this->assertMatchesRegularExpression('/^[@#$!]{64}$/', $randomString);
```

File: lib/public/Security/ISecureRandom.php [hunk 5]
```diff
@@ -46,22 +51,39 @@ interface ISecureRandom {
...
+	 * @param int $length Number of characters (must be > 0).
+	 * @param string $characters Optional list of unique, single-byte (ASCII) characters
+	 *        to use. Defaults to the CHAR_BASE64_RFC4648 alphabet. A custom set should contain at least 4
+	 *        characters, and must not contain duplicates or multibyte (non-ASCII) characters. It is strongly
+	 *        recommended to use predefined constants from ISecureRandom, which all meet the requirements.
+	 * @return string The randomly generated string.
+	 * @throws \LengthException If $length <= 0.
+	 * @throws \InvalidArgumentException if $characters contains non-ASCII characters, duplicates,
+	 *         or fewer than 4 unique characters.
 	 * @since 8.0.0
```
Comment (nickvergessen) [inline lib/public/Security/ISecureRandom.php:85]:
```suggestion
	 * @since 35.0.0 $characters has to be at least 4 chars long, non-AS… etc
	 * @since 8.0.0
```

Add a `@since` for your change to the rules of the method

File: tests/lib/Security/SecureRandomTest.php (hunk 1 above)
Comment (joshtrichards) [reply tests/lib/Security/SecureRandomTest.php]: Done

File: tests/lib/Security/SecureRandomTest.php (hunk 2 above)
Comment (joshtrichards) [reply tests/lib/Security/SecureRandomTest.php]: Done - apologies for the delay!

**Review by nickvergessen — APPROVED** [review]
